src/Rector/Architecture/RepositoryAsService/ServiceLocatorToDIRector.php

```diff
@@ -0,0 +1,129 @@
+<?php declare(strict_types=1);
+
+namespace Rector\Rector\Architecture\RepositoryAsService;
+
+use get_class;
+use Nette\Utils\Strings;
+use PhpParser\Node;
+use PhpParser\Node\Expr\ClassConstFetch;
+use PhpParser\Node\Name;
+use PhpParser\Node\Scalar\String_;
+use Rector\Builder\Class_\ClassPropertyCollector;
+use Rector\Contract\Bridge\RepositoryForDoctrineEntityProviderInterface;
+use Rector\Exception\Bridge\RectorProviderException;
+use Rector\Exception\ShouldNotHappenException;
+use Rector\Node\Attribute;
+use Rector\Node\PropertyFetchNodeFactory;
+use Rector\NodeAnalyzer\MethodCallAnalyzer;
+use Rector\Rector\AbstractRector;
+
+final class ServiceLocatorToDIRector extends AbstractRector
+{
+    /**
+     * @var MethodCallAnalyzer
+     */
+    private $methodCallAnalyzer;
+
+    /**
+     * @var PropertyFetchNodeFactory
+     */
+    private $propertyFetchNodeFactory;
+
+    /**
+     * @var RepositoryForDoctrineEntityProviderInterface
+     */
+    private $repositoryForDoctrineEntityProvider;
+
+    /**
+     * @var ClassPropertyCollector
+     */
+    private $classPropertyCollector;
+
+    public function __construct(
+        MethodCallAnalyzer $methodCallAnalyzer,
+        PropertyFetchNodeFactory $propertyFetchNodeFactory,
+        RepositoryForDoctrineEntityProviderInterface $repositoryForDoctrineEntityProvider,
+        ClassPropertyCollector $classPropertyCollector
+    ) {
+        $this->methodCallAnalyzer = $methodCallAnalyzer;
+        $this->propertyFetchNodeFactory = $propertyFetchNodeFactory;
+        $this->repositoryForDoctrineEntityProvider = $repositoryForDoctrineEntityProvider;
+        $this->classPropertyCollector = $classPropertyCollector;
+    }
+
+    public function isCandidate(Node $node): bool
+    {
+        if (! $this->methodCallAnalyzer->isMethod($node, 'getRepository')) {
+            return false;
+        }
+
+        $className = $node->getAttribute(Attribute::CLASS_NAME);
+
+        if ($className === null) {
+            return false;
+        }
+
+        return ! Strings::endsWith($node->getAttribute(Attribute::CLASS_NAME), 'Repository');
+    }
+
+    public function refactor(Node $node): ?Node
+    {
+        $this->classPropertyCollector->addPropertyForClass(
+            (string) $node->getAttribute(Attribute::CLASS_NAME),
+            [$this->repositoryFqn($node)],
+            $this->repositoryVariableName($node)
+        );
+
+        return $this->propertyFetchNodeFactory->createLocalWithPropertyName(
+            $this->repositoryVariableName($node)
```

Comment:
I wonter if this method doesn't suggest class decoupling.
It's connected to 4 private methods. They could be easier tested while decoupled

Reply:
I'm not sure what you mean.
Could you provide error or code?

Reply:
I see. You might not create service that would depend on interface and pass this array as an argument to its method.

Reply:
What about this way?